### ISSUE TITLE
UdpClient Performance Issue

### ISSUE THREAD

**Issue by szguoxz** [dotnet/runtime#110719]
This code is copied from UDPClient, The decision made here is really ODD.
Aparently, 
      tempRemoteEP has nothing to do with remoteEP, which is a bug here, client will receive message from all EP instead of specified EP.

      Then, simply because you didn't return a length, you decide to make a copy of the buffer, then your buffer is a new buffer, which would cause performance issue, and the worse, sometimes it's the same buffer you always use. If I assume it's always a new buffer, it will create bugs!

```c#
      public byte[] Receive([NotNull] ref IPEndPoint? remoteEP)
        {
            ThrowIfDisposed();

            // this is a fix due to the nature of the ReceiveFrom() call and the
            // ref parameter convention, we need to cast an IPEndPoint to it's base
            // class EndPoint and cast it back down to IPEndPoint. ugly but it works.
            EndPoint tempRemoteEP = _family == AddressFamily.InterNetwork ?
                IPEndPointStatics.Any :
                IPEndPointStatics.IPv6Any;

            int received = Client.ReceiveFrom(_buffer, MaxUDPSize, 0, ref tempRemoteEP);
            remoteEP = (IPEndPoint)tempRemoteEP;

            // because we don't return the actual length, we need to ensure the returned buffer
            // has the appropriate length.

            if (received < MaxUDPSize)
            {
                byte[] newBuffer = new byte[received];
                Buffer.BlockCopy(_buffer, 0, newBuffer, 0, received);
                return newBuffer;
            }
            return _buffer;
        }
```

**Comment by huoyaoyuan** [comment]
> tempRemoteEP has nothing to do with remoteEP, which is a bug here

This is incorrect. `remoteEP` is a `ref` variable, which may be changed out of the method. `tempRemoteEP` is a copy to new location and won't be changed.

> Then, simply because you didn't return a length, you decide to make a copy of the buffer

`UDPClient` isn't actively maintained or designed for performance. It doesn't provide convenience over `Socket`. `Socket` provides a lot of overloads of `RecieveFrom{Async}` to allow better control of buffer.

**Comment by szguoxz** [comment]
You don’t understand what I am talking about. I mean, in the code, tempEP has nothing to do with RemoteEP. But they should!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by wfurt** [comment]
With Socket.ReceiveFrom you can use pool allocator and return the buffer when it is processed by other thread. Socket api in general provides more control.

**Comment by huoyaoyuan** [comment]
How could we provide "recommendation against" for `UdpClient`, and potentially `TcpClient`? The api of `Socket` is a mixture, can there be a documentation to highlight APIs for typical TCP and UDP usage?

**Comment by liveans** [comment]
> How could we provide "recommendation against" for `UdpClient`, and potentially `TcpClient`? The api of `Socket` is a mixture, can there be a documentation to highlight APIs for typical TCP and UDP usage?

We have it for [TcpClient](https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/sockets/tcp-classes#finite-tcp-control-with-the-socket-class), but not for `UdpClient`.

**Comment by antonfirsov** [comment]
We should consider creating a similar docs for UDP. We definitely don't want to invest into `UdpClient` perf.

**Comment by GF-Huang** [comment]
We need something like this for `UdpClient`:
```cs
readonly struct record ImmutableIPEndPoint(IPAddress Address, int Port);
public async ValueTask<ImmutableIPEndPoint> ReceiveAsync(Memory<byte> buffer, CancellationToken token = default);
```

**Comment by huoyaoyuan** [comment]
Just use `Socket`:

```csharp
await serverSocket.ReceiveFromAsync(buffer, endPoint, cancellationToken)
```

The `IPEndPoint` instance can be reused.

This is not the right issue for discussion. `UdpClient` (and also `TcpClient`) were never designed for performance.

**Comment by GF-Huang** [comment]
> Just use `Socket`:
> 
> await serverSocket.ReceiveFromAsync(buffer, endPoint, cancellationToken)
> The `IPEndPoint` instance can be reused.
> 

I remember the `endPoint` parameter is not used for receive remote endpoint. If I recall and understand correctly

**Comment by wfurt** [comment]
If you care about perf, use the overload with SocketAddress @GF-Huang 
UdpClient was added long time ago as convenience API but I don't think it offers much value.  Especially if you care about perf & details. 

```c#
public int ReceiveFrom(Span<byte> buffer, System.Net.Sockets.SocketFlags socketFlags, System.Net.SocketAddress receivedAddress);
``` 
that does not allocate at all - unlike the case with EndPoint.
